atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/SnapshotStoreImplTest.java

```diff
@@ -0,0 +1,101 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.cache;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CellReference;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.watch.Sequence;
+import com.palantir.atlasdb.keyvalue.api.watch.StartTimestamp;
+import io.vavr.collection.HashMap;
+import io.vavr.collection.HashSet;
+import java.util.stream.Stream;
+import org.junit.Before;
+import org.junit.Test;
+
+public final class SnapshotStoreImplTest {
+    private static final Sequence SEQUENCE_1 = Sequence.of(1337L);
+    private static final Sequence SEQUENCE_2 = Sequence.of(8284L);
+    private static final StartTimestamp TIMESTAMP_1 = StartTimestamp.of(42L);
+    private static final StartTimestamp TIMESTAMP_2 = StartTimestamp.of(31415925635L);
+    private static final StartTimestamp TIMESTAMP_3 = StartTimestamp.of(404L);
+    private static final StartTimestamp TIMESTAMP_4 = StartTimestamp.of(10110101L);
+    private static final ValueCacheSnapshot SNAPSHOT_1 = ValueCacheSnapshotImpl.of(HashMap.empty(), HashSet.empty());
+    private static final ValueCacheSnapshot SNAPSHOT_2 = ValueCacheSnapshotImpl.of(
+            HashMap.<CellReference, CacheEntry>empty()
+                    .put(
+                            CellReference.of(
+                                    TableReference.createFromFullyQualifiedName("t.table"),
+                                    Cell.create(new byte[] {1}, new byte[] {1})),
+                            CacheEntry.locked()),
+            HashSet.empty());
+    private SnapshotStore snapshotStore;
+
+    @Before
+    public void before() {
+        snapshotStore = new SnapshotStoreImpl();
+    }
+
+    @Test
+    public void singleSnapshotStoredForMultipleTimestamps() {
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_1, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_2, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_3, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_2, TIMESTAMP_4, SNAPSHOT_2);
+
+        assertSnapshotsEqualForTimestamp(SNAPSHOT_1, TIMESTAMP_1, TIMESTAMP_2, TIMESTAMP_3);
+        assertSnapshotsEqualForTimestamp(SNAPSHOT_2, TIMESTAMP_4);
+    }
+
+    @Test
+    public void snapshotsNotOverwrittenForSameSequence() {
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_1, SNAPSHOT_1);
+        snapshotStore.storeSnapshot(SEQUENCE_1, TIMESTAMP_2, SNAPSHOT_2);
```

Comment:
Should this even fail silently or throw?

Reply:
You're right, probably should fail loudly.